### PR TITLE
WT-13333 Process logged tables only when checkpoint cleanup is configured to be aggressive

### PR DESCRIPTION
**Pull request by etienneptl**
The connection configuration string `checkpoint_cleanup=[method=reclaim_space]` can be used to process logged tables during checkpoint cleanup. The associated connection flag has been renamed from `WT_CONN_CKPT_CLEANUP_SKIP_INT` to `WT_CONN_CKPT_CLEANUP_RECLAIM_SPACE`.

### CODE REVIEW

**Comment by etienneptl** [comment]
Would be great to come up with a test for this new feature.

File: src/btree/bt_sync_obsolete.c [hunk 1]
```diff
@@ -406,16 +406,19 @@ __checkpoint_cleanup_page_skip(
      * FIXME: Read internal pages from non-logged tables when the remove/truncate
      * operation is performed using no timestamp.
      */
+    if (addr.type == WT_ADDR_LEAF_NO)
+        *skipp = true;
+    else if (addr.ta.newest_stop_durable_ts == WT_TS_NONE)
+        /* Only process logged tables when checkpoint cleanup is configured to be aggressive. */
+        *skipp = !(F_ISSET(S2C(session), WT_CONN_CKPT_CLEANUP_RECLAIM_SPACE) &&
```
Comment (kommiharibabu) [inline src/btree/bt_sync_obsolete.c:413]: This condition is wrong.

Method - None.
1. Logged table - (true & true) - skip.
2. Non-logged table - (true & false) - don't skip (Actually it should skip)

Method - reclaim space
1. Logged table - (false & true) - Do not skip
2. Non-logged table - (true & false) - Do not skip (Actually it should skip)

Comment (kommiharibabu) [reply src/btree/bt_sync_obsolete.c:413]: Sorry, the condition is correct. I thought the `!` was only for the flag and not for whole condition.

Comment (etienneptl) [reply src/btree/bt_sync_obsolete.c:413]: @kommiharibabu, I am happy to change it to:
```
        *skipp = !F_ISSET(S2C(session), WT_CONN_CKPT_CLEANUP_RECLAIM_SPACE) ||
          !F_ISSET(S2BT(session), WT_BTREE_LOGGED);
```
Would that be easier to read?

**Review by kommiharibabu — APPROVED** [review]
LGTM

File: src/btree/bt_sync_obsolete.c [hunk 2]
```diff
@@ -406,16 +406,19 @@ __checkpoint_cleanup_page_skip(
...
+          F_ISSET(S2BT(session), WT_BTREE_LOGGED));
```
Comment (marcbutler) [inline src/btree/bt_sync_obsolete.c:414]: Sorry but I'm not following the correlation comment and the code.
This code reads to me:
skip = true if reclaim == false and logged == false
but the comment appears to indicate the opposite.

File: src/btree/bt_sync_obsolete.c (hunk 1 above)
Comment (kommiharibabu) [reply src/btree/bt_sync_obsolete.c:413]: Yes, it may be easier to read the code.

File: src/btree/bt_sync_obsolete.c (hunk 2 above)
Comment (etienneptl) [reply src/btree/bt_sync_obsolete.c:414]: Yes, if we are not operating in aggressive mode (reclaim == false), we want to skip.

Comment (etienneptl) [reply src/btree/bt_sync_obsolete.c:414]: It looks like the condition is hard to read, I am happy to flip it and remove the `!` at the very start as suggested [here](https://github.com/wiredtiger/wiredtiger/pull/10918#discussion_r1693910372).

Comment (marcbutler) [reply src/btree/bt_sync_obsolete.c:414]: Would you consider the following:
```
/* top of function */
const unsigned int AGGRESSIVE_MODE = (WT_CONN_CKPT_CLEANUP_RECLAIM_SPACE|WT_BTREE_LOGGED);

*skipp = (F_MASK(S2C(session), AGGRESSIVE_MODE) != AGGRESSIVE_MODE);
```

Comment (etienneptl) [reply src/btree/bt_sync_obsolete.c:414]: I'd like to keep the notion of aggressive mode and logging separate. I can imagine that we may do different things unrelated to logged tables if the aggressive mode is enabled.

Comment (etienneptl) [reply src/btree/bt_sync_obsolete.c]: Updated condition in 204385f31b5ff05a56e997fbfedf199a40c046a6

File: src/btree/bt_sync_obsolete.c (hunk 1 above)
Comment (etienneptl) [reply src/btree/bt_sync_obsolete.c]: Updated in 204385f31b5ff05a56e997fbfedf199a40c046a6

**Review by marcbutler — APPROVED** [review]
LGTM

**Review by quchenhao — APPROVED** [review]
lgtm

File: test/suite/test_cc01.py [hunk 3]
```diff
@@ -77,6 +77,14 @@ def check(self, check_value, uri, nrows, read_ts):
         session.rollback_transaction()
         self.assertEqual(count, nrows)
 
+    def populate(self, uri, start_key, num_keys, value):
```
Comment (quchenhao) [inline test/suite/test_cc01.py:80]: Why you need to add this?

File: test/suite/test_cc08.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+
+from test_cc01 import test_cc_base
+from wiredtiger import stat
+from wtscenario import make_scenarios
+
+# test_cc08.py
+# Verify checkpoint cleanup cleans up logged tabled when configured in aggressive mode.
```
Comment (quchenhao) [inline test/suite/test_cc08.py:34]: typo tabled

File: test/suite/test_cc08.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+        # Write everything to disk.
+        self.session.checkpoint()
+
+        # Restart and specify the checkpoint cleanup's strategy.
+        self.reopen_conn(config=f'{self.conn_config},{self.cc_config}')
+
+        # Open the table as we need the dhandle to be open for checkpoint cleanup to process the
+        # table.
+        _ = self.session.open_cursor(uri, None, None)
```
Comment (quchenhao) [inline test/suite/test_cc08.py:64]: _ = is for what?

File: dist/stat_data.py [hunk 6]
```diff
@@ -1087,6 +1087,7 @@ def __init__(self, name, desc, flags=''):
     CheckpointStat('checkpoint_cleanup_pages_evict', 'pages added for eviction during checkpoint cleanup'),
     CheckpointStat('checkpoint_cleanup_pages_obsolete_tw', 'pages dirtied due to obsolete time window by checkpoint cleanup'),
     CheckpointStat('checkpoint_cleanup_pages_removed', 'pages removed during checkpoint cleanup'),
+    CheckpointStat('checkpoint_cleanup_pages_selected_reclaim_space', 'pages selected during checkpoint cleanup (reclaim_space)'),
```
Comment (kommiharibabu) [inline dist/stat_data.py:1090]: We will read the page into the cache without skipping it. How about mentioning `read` in the stat name? 
`pages read into the cache ...`?

File: test/suite/test_cc01.py (hunk 3 above)
Comment (etienneptl) [reply test/suite/test_cc01.py:80]: It's part of the base class, we can now use `self.populate()` in any `cc` test which I am doing in this PR.

File: test/suite/test_cc08.py (hunk 5 above)
Comment (etienneptl) [reply test/suite/test_cc08.py:64]: Unused returned value.

File: test/suite/test_cc08.py (hunk 4 above)
Comment (etienneptl) [reply test/suite/test_cc08.py]: Thank you, see 46264f29cffbe72c0afe256dce71cfbfdfe62d7d

File: dist/stat_data.py (hunk 6 above)
Comment (etienneptl) [reply dist/stat_data.py]: Sure, see 5a889d064b2828d725ed4ac8f9d3aa0b11f3e972

**Comment by marcbutler** [comment]
I appreciate it is not required for review but from my perspective a description would be helpful.

**Comment by etienneptl** [comment]
> I appreciate it is not required for review but from my perspective a description would be helpful.

Done! Thank you for pointing it out.

**Review by quchenhao — APPROVED** [review]
lgtm

**Review by bvpvamsikrishna — APPROVED** [review]
lgtm

**Pull request merged**